### DISCUSSION TITLE
How to Use PostGIS to Convert Address to Latitude and Longitude in PostgreSQL

### DISCUSSION THREAD

**Discussion by organicnz** [supabase/supabase#14163] [Questions] [answered]
Hi everyone,

I want to convert the `post_address` column with `address` format 4621 Santa Monica Blvd Los Angeles, CA 90029 to the data type `long` `lat` in the `locations` column of the `posts` table in Postgresql. However, the SQL query drops an error `Failed to run SQL query: function geocode(text) doesn't exist`. Is there a workaround or best practice for a similar implementation? :) 

Thank you :)

SQL query:
```CREATE EXTENSION IF NOT EXISTS postgis;
CREATE EXTENSION IF NOT EXISTS postgis_tiger_geocoder;

CREATE OR REPLACE FUNCTION update_locations()
RETURNS TRIGGER AS $$
BEGIN
  UPDATE posts SET locations = ST_SetSRID(ST_MakePoint((result::geometry).x, (result::geometry).y), 4326)
  FROM (SELECT NEW.id, geocode(NEW.post_address) AS result) AS geocoded
  WHERE posts.id = geocoded.id;
  RETURN NEW;
END;
$$ LANGUAGE plpgsql;

CREATE TRIGGER update_posts_locations
AFTER INSERT OR UPDATE OF post_address ON posts
FOR EACH ROW
EXECUTE FUNCTION update_locations();
```

![image](https://user-images.githubusercontent.com/4765090/236542984-838a9793-53d9-4399-bd8a-cef000a967de.png)

**Comment by GaryAustin1** [comment]
You may need to set the search_path in your trigger to include extensions or the schema the function is installed in. 
Most examples show geocode('text address', 1)   Where 1 is the limit of results to return.   I think it is optional but not sure.

**Comment by organicnz** [comment]
Thank you very much Gary,

Not sure, if it's the right modification :)

```
-- install the PostGIS extension
CREATE EXTENSION IF NOT EXISTS postgis;

-- create the trigger function to update locations
CREATE OR REPLACE FUNCTION update_locations()
RETURNS TRIGGER AS $$
BEGIN
  NEW.locations := ST_SetSRID(geocode(NEW.post_address, 1), 4326);
  RETURN NEW;
END;
$$ LANGUAGE plpgsql
SET search_path = public, postgis_tiger_geocoder;

-- create the trigger to update locations on insert or update of post_address
CREATE TRIGGER update_posts_locations
BEFORE INSERT OR UPDATE OF post_address
ON posts
FOR EACH ROW
EXECUTE FUNCTION update_locations();
```

`SELECT PostGIS_full_version();`

`POSTGIS="3.1.4 ded6c34" [EXTENSION] PGSQL="140" GEOS="3.8.0-CAPI-1.13.1 " SFCGAL="1.3.10" PROJ="6.3.1" GDAL="GDAL 3.0.4, released 2020/01/28" LIBXML="2.9.10" LIBJSON="0.13.1" TOPOLOGY RASTER`


![image](https://user-images.githubusercontent.com/4765090/236616616-7ad9eca0-d8ab-4a3c-83e1-a28a709d75db.png)

**Reply by GaryAustin1** [reply]
Did you try tiger for the schema?

**Reply by organicnz** [reply]
That's all I've tried so far. Although I could try asking ChatGPT for a specific Tiger trigger function :)

**Reply by GaryAustin1** [reply]
I get things to "run" by adding set search_path = public,extensions, tiger;  Had to install extension fuzzystrmatch also.
![image](https://user-images.githubusercontent.com/54564956/236639441-67ad3606-5dc3-44fd-99f5-99eb26de7d3f.png)

But I get back null result.  I'm really not sure how postgis_tiger_decoder gets set up in Supabase.  I believe you need a data table, I'm not sure that gets loaded automatically, or you have to load the data geocode needs for your area.

**Reply by organicnz** [reply]
Thank you, Gary. Yeah, these extensions were already enabled before trigger functions were set up :)

![image](https://user-images.githubusercontent.com/4765090/236672700-73cf808d-2809-45a5-961d-899eef798c74.png)

![image](https://user-images.githubusercontent.com/4765090/236672738-579c63ca-00b9-495b-af58-5e183efe625e.png)

**Reply by scottwyn13** [reply]
Did you find a resolution to this? I am trying to accomplish the same thing but get the same "function geocode(text, integer) does not exist" error.

**Reply by organicnz** [reply]
Sorry for the late one. Nope, I'm encountering the same error with the other trigger function :(

<img width="563" alt="Screenshot 2024-06-11 at 12 42 33" src="https://github.com/supabase/supabase/assets/4765090/666b7daf-c01a-4dcd-ba69-b308b6821d08">

**Reply by organicnz** [reply, marked as answer]
I've found a way to use edge functions to specifically handle coordinate updates, which works really well with the trigger function.

https://github.com/orgs/supabase/discussions/27167#discussioncomment-10003194
